### PR TITLE
Add some circle intersection and tangent utilities to QgsGeometryUtils

### PR DESCRIPTION
**Pull request by nyalldawson**
Adds some more useful circle-related methods to QgsGeometryUtils

### CODE REVIEW

**Comment by lbartoletti** [comment]
Well done. I prepared some improvements to QgsCircle like circle intersection :)

Just an open question. Why add these methods in QgsGeometryUtils instead of QgsCircle?

Thanks.

**Comment by nyalldawson** [comment]
@lbartoletti

I was hoping you'd review :) 

I figured adding them to QgsGeometryUtils as a start was a good move, because that means they can be used without constructing a QgsCircle (and accordingly be slightly more efficient if an existing QgsCircle isn't available). But I'd be in favour of adding these functions to QgsCircle too and calling the QgsGeometryUtils method internally.

**Comment by lbartoletti** [comment]
@nyalldawson OK. LGTM :)

When you will add these methods in QgsCircle, please add (a little) Z support. For example, an option to enable a 2D intersection (default method) or if circles have the same Z add the Z in intersection(s) point(s) (special cases).

Thanks!

File: src/core/geometry/qgsgeometryutils.h [hunk 1]
```diff
@@ -154,6 +154,53 @@ class CORE_EXPORT QgsGeometryUtils
...
+     *
+     * If found, the intersection points will be stored in \a intersection1 and \a intersection2.
+     *
+     * \returns number of intersection points found.
+     *
+     * \since QGIS 3.2
+     */
+    static int circleCircleIntersections( QgsPointXY center1, double radius1,
+                                          QgsPointXY center2, double radius2,
+                                          QgsPointXY &intersection1, QgsPointXY &intersection2 );
```
Comment (3nids) [inline src/core/geometry/qgsgeometryutils.h]: missing `SIP_OUT` for intersections

File: src/core/geometry/qgsgeometryutils.h [hunk 2]
```diff
@@ -154,6 +154,53 @@ class CORE_EXPORT QgsGeometryUtils
...
+     * and the second line is described by the points stored in \a line2P1
+     * and \a line2P2.
+     *
+     * Returns the number of tangents (either 0 or 2).
+     *
+     * \since QGIS 3.2
+     */
+    static int circleCircleOuterTangents(
+      const QgsPointXY &center1, double radius1, const QgsPointXY &center2, double radius2,
+      QgsPointXY &line1P1, QgsPointXY &line1P2,
```
Comment (3nids) [inline src/core/geometry/qgsgeometryutils.h]: same

File: src/core/geometry/qgsgeometryutils.h [hunk 3]
```diff
@@ -154,6 +154,53 @@ class CORE_EXPORT QgsGeometryUtils
...
+
+    /**
+     * Calculates the tangent points between the circle with the specified \a center and \a radius
+     * and the point \a p.
+     *
+     * If found, the tangent points will be stored in \a pt1 and \a pt2.
+     *
+     * \since QGIS 3.2
+     */
+    static bool tangentPointAndCircle( const QgsPointXY &center, double radius,
```
Comment (3nids) [inline src/core/geometry/qgsgeometryutils.h]: same

**Review by 3nids — COMMENTED** [review]
this misses some SIP_OUT

File: src/core/geometry/qgsgeometryutils.h (hunk 1 above)
Comment (nyalldawson) [reply src/core/geometry/qgsgeometryutils.h]: Fixed (x3)

**Comment by agiudiceandrea** [comment]
Hi @nyalldawson, please see issue report https://github.com/qgis/QGIS/issues/54322 about circleCircleIntersections.

**Pull request merged**